crates/sui-graphql-rpc/src/types/event/mod.rs

```diff
@@ -158,64 +115,67 @@ impl Event {
-                                transactions::dsl::transactions
-                                    .select(transactions::dsl::tx_sequence_number)
-                                    .filter(
-                                        transactions::dsl::transaction_digest.eq(digest.to_vec()),
-                                    )
-                                    .single_value(),
-                            ),
-                        )
-                    }
-
-                    if let Some(module) = &filter.emitting_module {
-                        query = module.apply(query, events::dsl::package, events::dsl::module);
-                    }
-
-                    if let Some(type_) = &filter.event_type {
-                        query = type_.apply(
-                            query,
-                            events::dsl::event_type,
-                            events::dsl::event_type_package,
-                            events::dsl::event_type_module,
-                            events::dsl::event_type_name,
-                        );
-                    }
-
-                    query
-                })
+                let tx_hi: i64 = conn.first(move || {
+                    dsl::checkpoints.select(dsl::network_total_transactions)
+                        .filter(dsl::sequence_number.eq(checkpoint_viewed_at as i64))
+                })?;
+
+                query = add_bounds(query, &filter, &page, (tx_hi - 1) as u64);
+
+                let (prev, next, results) =
+                    page.paginate_raw_query::<EvLookup>(conn, checkpoint_viewed_at, query)?;
+
+                let ev_lookups = results
+                    .into_iter()
+                    .map(|x| (x.tx, x.ev))
+                    .collect::<Vec<(i64, i64)>>();
+
+                if ev_lookups.is_empty() {
+                    return Ok::<_, diesel::result::Error>((prev, next, vec![]));
+                }
+
+                // Unlike a multi-get on a single column which can be serviced by a query `IN
+                // (...)`, because events have a composite primary key, the query planner tends
+                // to perform a sequential scan when given a list of tuples to lookup. A query
+                // using `UNION ALL` allows us to leverage the index on the composite key.
+                let mut events: Vec<StoredEvent> = conn.results(move || {
+                    // Diesel's DSL does not current support chained `UNION ALL`, so we have to turn
+                    // to `RawQuery` here.
+                    let query_string = ev_lookups.iter()
+                    .map(|&(tx, ev)| {
+                        format!("SELECT * FROM events WHERE tx_sequence_number = {} AND event_sequence_number = {}", tx, ev)
+                    })
+                    .collect::<Vec<String>>()
+                    .join(" UNION ALL ");
+
+                    query!(query_string).into_boxed()
```

Comment:
Does it work to chain `or_filter` with `tx_sequence_number.eq(tx).and(event_sequence_number.eq(ev))`? Asking because it would be helpful for type-safety to use the non-raw query to fetch the event contents.

Reply:
It's possible but will `or`s of `and`s work performance-wise? Will postgres be able to take advantage of the index on the composite key?

Reply:
Sorry, I should have clarified, the performance question was what I was asking as well: If we construct an OR of ANDs, will postgres' query planner be kind to us? I suspect it might be (it's not meaningfully different to doing the union).

Reply:
If we use `OR` or `IN`, the query planner will choose to do a sequential scan on events partitions, which is what drove me to use `UNION ALL`